cedar-lean/Cedar/Thm/Lemmas/Typechecker.lean

```diff
@@ -0,0 +1,167 @@
+This file contains useful definitions and lemmas about the `Typechecker` functions.
+
+todo: fill in `sorry`s. Some invariants may need to be adjusted. The current
+definitions are an informed guess based on the corresponding Dafny proof.
+-/
+
+namespace Cedar.Thm
+
+open Cedar.Data
+open Cedar.Spec
+open Cedar.Validation
+
+def InstanceOfBoolType (b : Bool) (bty : BoolType) : Prop :=
+  match (b, bty) with
+  | (true, .tt) => True
+  | (false, .ff) => True
+  | (_, .anyBool) => True
+  | _ => False
+
+def InstanceOfEntityType (e : EntityUID) (ety: EntityType) : Prop :=
+  ety = e.ty
+
+def InstanceOfExtType (e : Ext) (ety : ExtType) : Prop :=
+  match (e, ety) with
+  | (.decimal _, .decimal) => True
+  | (.ipaddr _, .ipAddr) => True
+  | _ => False
+
+inductive InstanceOfType : Value → CedarType → Prop :=
+  | instance_of_bool (b : Bool) (bty : BoolType) : InstanceOfBoolType b bty →
+      InstanceOfType (.prim (.bool b)) (.bool bty)
+  | instance_of_int : InstanceOfType (.prim (.int _)) .int
+  | instance_of_string : InstanceOfType (.prim (.string _)) .string
+  | instance_of_entity (e : EntityUID) (ety: EntityType) : InstanceOfEntityType e ety →
+      InstanceOfType (.prim (.entityUID e)) (.entity ety)
+  | instance_of_set (s : Set Value) (ty : CedarType) : (forall v, v ∈ s → InstanceOfType v ty) →
+      InstanceOfType (.set s) (.set ty)
+  | instance_of_record (r : Map Attr Value) (rty : RecordType) :
+      -- if an attribute is present, then it has the expected type
+      (∀ (k : Attr) (v : Value) (qty : QualifiedType),
+        rty.find? k = some qty → r.find? k = some v → InstanceOfType v qty.getType) →
+      -- required attributes are present
+      (∀ (k : Attr) (qty : QualifiedType), rty.find? k = some qty → qty.isRequired → r.contains k) →
+      InstanceOfType (.record r) (.record rty)
+  | instance_of_ext (e : Ext) (ety : ExtType) : InstanceOfExtType e ety →
+      InstanceOfType (.ext e) (.ext ety)
+
+def InstanceOfRequestType (request : Request) (reqty : RequestType) : Prop :=
+  InstanceOfEntityType request.principal reqty.principal ∧
+  request.action = reqty.action ∧
+  InstanceOfEntityType request.resource reqty.resource ∧
+  InstanceOfType request.context (.record reqty.context)
+
+/--
+For every entity in the store,
+1. The entity's type is defined in the type store.
+2. The entity's attributes match the attribute types indicated in the type store.
+3. The entity's ancestors' types are defined in the type store, and the ancestor
+   relationship is consistent with the descendants information in the type store.
+-/
+def InstanceOfEntityTypeStore (entities : Entities) (ets: EntityTypeStore) : Prop :=
+  ∀ uid data, entities.find? uid = some data →
+    ∃ attrTys descendantTys, ets.find? uid.ty = some (attrTys, descendantTys) ∧
+      InstanceOfType data.attrs (.record attrTys) ∧
+        ∀ ancestor, ancestor ∈ data.ancestors →
+          ∃ descendants, ets.find? ancestor.ty = some descendants ∧ uid.ty ∈ descendantTys
```

Comment:
This doesn't look right to me.  `EntityTypeStore` maps an EntityType to the set of its ancestor types, rather than its descendent types.  See [here](https://github.com/cedar-policy/cedar-spec/blob/0b0cceafabdbcb7313d753438209ea6bbf4692c7/cedar-lean/Cedar/Validation/Types.lean#L77).

IIRC, the Dafny formalization had this flipped compared to how it's defined in the `EntityStore`, and in the Lean formalization, we decided to keep the direction consistent (always ancestors).

Reply:
Ah, I didn't look closely enough at the definition of `EntityTypeStore.descendentOf`. The formalization in Dafny matches the Rust code (the store tracks ancestors while the schema tracks descendants). Do we want to match the Rust code to simplify DRT, or switch to the more consistent formalization?

Reply:
My vote would be for the more consistent formalization :)

Reply:
Updated to the more consistent formalization. Please re-review my latest commit when you get a chance